### PR TITLE
New request handling method

### PR DESCRIPTION
**Pull request by omar-mohamed-khallaf**

### CODE REVIEW

**Comment by omar-mohamed-khallaf** [comment]
**Sending 2000 concurrent requests**

```
This is ApacheBench, Version 2.3 <$Revision: 1903618 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient).....done


Server Software:        
Server Hostname:        localhost
Server Port:            8080

Document Path:          /api/login
Document Length:        20 bytes

Concurrency Level:      2000
Time taken for tests:   0.287 seconds
Complete requests:      2000
Failed requests:        0
Total transferred:      182000 bytes
Total body sent:        372000
HTML transferred:       40000 bytes
Requests per second:    6958.36 [#/sec] (mean)
Time per request:       287.424 [ms] (mean)
Time per request:       0.144 [ms] (mean, across all concurrent requests)
Transfer rate:          618.37 [Kbytes/sec] received
                        1263.92 kb/s sent
                        1882.29 kb/s total

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0   54   6.3     53      66
Processing:    38  126  41.3    112     182
Waiting:        5  125  40.7    112     178
Total:         66  180  35.6    167     232

Percentage of the requests served within a certain time (ms)
  50%    167
  66%    221
  75%    222
  80%    222
  90%    222
  95%    223
  98%    223
  99%    223
 100%    232 (longest request)
```

**Sending 100000 request**

```
This is ApacheBench, Version 2.3 <$Revision: 1903618 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient).....done


Server Software:        
Server Hostname:        localhost
Server Port:            8080

Document Path:          /api/login
Document Length:        20 bytes

Concurrency Level:      16
Time taken for tests:   6.026 seconds
Complete requests:      100000
Failed requests:        0
Total transferred:      9100000 bytes
Total body sent:        18600000
HTML transferred:       2000000 bytes
Requests per second:    16595.05 [#/sec] (mean)
Time per request:       0.964 [ms] (mean)
Time per request:       0.060 [ms] (mean, across all concurrent requests)
Transfer rate:          1474.75 [Kbytes/sec] received
                        3014.33 kb/s sent
                        4489.09 kb/s total

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.2      0       3
Processing:     0    0   0.2      0       9
Waiting:        0    0   0.2      0       9
Total:          1    1   0.1      1       9

Percentage of the requests served within a certain time (ms)
  50%      1
  66%      1
  75%      1
  80%      1
  90%      1
  95%      1
  98%      1
  99%      1
 100%      9 (longest request)
```

File: src/test/SimpleLogin.java [hunk 1]
```diff
@@ -1,62 +1,47 @@
...
             if (!(LoginInfo.containsKey("username") && LoginInfo.containsKey("password"))) {
-                response.put("body", "Invalid Request.".getBytes());
-                response.put("code", HTTPCode.BAD_REQUEST);
+                httpResponse.setBody("{\"success\": \"false\"}");
             } else {
                 if (LoginInfo.get("username").equals(username) && LoginInfo.get("password").equals(password)) {
-                    response.put("body", JSON.HMQ(new HashMap<>() {{
-                        put("status", "success");
-                    }}).getBytes());
+                    httpResponse.setBody("{\"success\": \"true\"}");
```
Comment (Zelakolase) [inline src/test/SimpleLogin.java]: Why don't you use JSON class?

File: src/test/SimpleLogin.java [hunk 2]
```diff
@@ -1,62 +1,47 @@
...
-        response.put("CustomHeaders", new HashMap<String, String>() {{
-            put("RandomInteger", String.valueOf(new Random().nextInt(10)));
-        }});
-        return response;
+        return httpResponse;
+    }
+
+    public static void main(String[] args) throws Exception {
+
+        Server server = new Server(SimpleLogin::handle);
```
Comment (Zelakolase) [inline src/test/SimpleLogin.java:75]: LGTM

File: src/test/ServeFiles.java [hunk 3]
```diff
@@ -1,33 +1,36 @@
...
+
+public class ServeFiles {
+    private static HttpResponse handle(HttpRequest httpRequest) {
+        HttpResponse httpResponse = new HttpResponse();
+        String path = httpRequest.getPath().replaceFirst("/", "www/");
+        try (BufferedInputStream f = new BufferedInputStream(new FileInputStream(path))) {
+            httpResponse.setHttpStatusCode(HttpStatusCode.OK);
+            httpResponse.setHttpContentType(HttpContentType.TEXT_PLAIN);
+            byte[] fileContents = f.readAllBytes();
+            httpResponse.setBody(new String(fileContents, 0, fileContents.length, StandardCharsets.UTF_8));
```
Comment (Zelakolase) [inline src/test/ServeFiles.java]: Why did you use String wrapper? What if it is a raw file (eg. mp4)?

Comment (Zelakolase) [reply src/test/ServeFiles.java]: Please add JKS comment, in the SimpleLogin class as well

File: src/server/Server.java [hunk 4]
```diff
@@ -159,46 +145,24 @@ private SSLContext getSSLContext(Path keyStorePath, char[] keyStorePass, String
         return sslContext;
     }
 
-    public HashMap<String, String> pathFiltration(HashMap<String, String> Headers) {
-        String path = "./" + WWWDir + PathFilter.filter(Headers.get("path"));
```
Comment (Zelakolase) [inline src/server/Server.java:163]: PathFilter is used to mitigate LFI attacks

File: src/server/Server.java [hunk 5]
```diff
@@ -9,92 +15,72 @@
...
+                s.setReceiveBufferSize(maxRequestSizeBytes);
+                s.setSendBufferSize(maxRequestSizeBytes);
+                s.setSoTimeout(60000);
 
                 try {
                     poolExecutor.execute(new Engine(s));
                 } catch (RejectedExecutionException ignore) {
                     log.i("concurrent connections exceed the configured maximum");
-                    Network.write(new BufferedOutputStream(s.getOutputStream()), new byte[]{}, "", HTTPCode.SERVICE_UNAVAILABLE, false, new HashMap<>(), false);
+                    Network.write(new BufferedOutputStream(s.getOutputStream()), new HttpResponse("", HttpStatusCode.SERVICE_UNAVAILABLE, HttpContentType.TEXT_PLAIN), false);
```
Comment (Zelakolase) [inline src/server/Server.java]: HttpContentType.PLAIN_TEXT is more intuitive

File: src/server/Server.java [hunk 6]
```diff
@@ -9,92 +15,72 @@
...
                 Socket s = serverSocket.accept();
-                s.setKeepAlive(false);
+                s.setKeepAlive(true);
                 s.setTcpNoDelay(true);
-                s.setReceiveBufferSize(BufferSize);
-                s.setSendBufferSize(BufferSize);
-                s.setSoTimeout(SOTimeout);
+                s.setReceiveBufferSize(maxRequestSizeBytes);
+                s.setSendBufferSize(maxRequestSizeBytes);
+                s.setSoTimeout(60000);
```
Comment (Zelakolase) [inline src/server/Server.java]: 60k should be default, but the user may want to change it, this change contradicts high customizability

File: src/server/Server.java [hunk 7]
```diff
@@ -9,92 +15,72 @@
...
+        throwable.printStackTrace(pw);
+        return sw.getBuffer().toString();
     }
 
     private void mainLoop(ServerSocket serverSocket, ThreadPoolExecutor poolExecutor) {
         while (true) {
             try {
                 Socket s = serverSocket.accept();
-                s.setKeepAlive(false);
+                s.setKeepAlive(true);
```
Comment (Zelakolase) [inline src/server/Server.java]: Is Slowloris mitigated? Also, the user should customize it.

File: src/server/Server.java [hunk 8]
```diff
@@ -9,92 +15,72 @@
...
 
-    public void setWWWDirectory(String in) {
-        WWWDir = in;
+    public Server(int port, int maxConcurrentRequests, int maxRequestSizeBytes, int maxResponseSizeBytes, boolean useGzip, Function<HttpRequest, HttpResponse> handler) {
+        this(port, maxConcurrentRequests, maxRequestSizeBytes, useGzip, "www", "etc/MIME.db", handler);
     }
 
-    public void setPort(int in) {
-        Port = in;
+    public Server(int port, int maxConcurrentRequests, int maxRequestSizeBytes, boolean useGzip, String rootDir, String mimeFile, Function<HttpRequest, HttpResponse> handler) {
```
Comment (Zelakolase) [inline src/server/Server.java]: Setters are better than constructors

File: src/server/Server.java [hunk 9]
```diff
@@ -9,92 +15,72 @@
...
+    private final Function<HttpRequest, HttpResponse> handler;
 
-    public void setMaximumRequestSizeInKB(int in) {
-        MaxRequestSizeKB = in;
-    }
 
-    public void setBacklog(int in) {
-        backlog = in;
+    public Server(Function<HttpRequest, HttpResponse> handler) {
+        this(8080, false, handler);
```
Comment (Zelakolase) [inline src/server/Server.java]: port 80

File: src/lib/Network.java [hunk 10]
```diff
@@ -24,79 +30,63 @@ public static byte[] compress(byte[] data) throws IOException {
...
-                if (counter < MAX_REQ_SIZE) {
-                    Reply.write(dIS.read());
-                    counter++;
-                } else break;
-            } while (dIS.available() > 0);
-        } catch (Exception ignored) {
+    public static String read(BufferedInputStream inputStream, int maxReqSize) throws IOException {
+        byte[] buffer = new byte[maxReqSize];
+        if (inputStream.read(buffer, 0, maxReqSize) == maxReqSize) {
+            throw new IOException("Request to big");
```
Comment (Zelakolase) [inline src/lib/Network.java]: typo

File: src/lib/Network.java [hunk 11]
```diff
@@ -24,79 +30,63 @@ public static byte[] compress(byte[] data) throws IOException {
...
-            do {
-                if (counter < MAX_REQ_SIZE) {
-                    Reply.write(dIS.read());
-                    counter++;
-                } else break;
-            } while (dIS.available() > 0);
-        } catch (Exception ignored) {
+    public static String read(BufferedInputStream inputStream, int maxReqSize) throws IOException {
+        byte[] buffer = new byte[maxReqSize];
+        if (inputStream.read(buffer, 0, maxReqSize) == maxReqSize) {
```
Comment (Zelakolase) [inline src/lib/Network.java]: The read function might not be able to report an accurate number. Moreover, if the remaining reqSize is x and the request is larger, we should read the x amount only. TLS Records handling is a bit tricky.

File: src/test/SimpleLogin.java (hunk 1 above)
Comment (Zelakolase) [reply src/test/SimpleLogin.java]: Please add an example for a custom header that changes every request, to clarify the impl.

**Review by Zelakolase — DISMISSED** [review]
Sorry for the 13 comments haha

**Comment by Zelakolase** [comment]
The merging will be late cuz I will make a README file.

File: src/server/Server.java (hunk 4 above)
Comment (Zelakolase) [reply src/server/Server.java:163]: Please reverse the deletion and use the function in ServeFiles class

File: src/test/SimpleLogin.java (hunk 1 above)
Comment (omar-mohamed-khallaf) [reply src/test/SimpleLogin.java]: > Why don't you use JSON class?

Just an example.

Comment (omar-mohamed-khallaf) [reply src/test/SimpleLogin.java]: > Please add an example for a custom header that changes every request, to clarify the impl.

```java
httpResponse.addHeader("key", "value");
```

File: src/server/Server.java (hunk 5 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java]: You're right but I wrote it to follow the pattern defined by the standard `type/subtype => text/plain => TEXT_PLAIN`

File: src/server/Server.java (hunk 9 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java]: Requires root access while developing and anyways the user can change it when creating the server, it's just the default port.

File: src/lib/Network.java (hunk 10 above)
Comment (omar-mohamed-khallaf) [reply src/lib/Network.java]: ?

File: src/lib/Network.java (hunk 11 above)
Comment (omar-mohamed-khallaf) [reply src/lib/Network.java]: The `read` function is guaranteed to read until

- The specified number of bytes have been read,
- The read method of the underlying stream returns -1, indicating end-of-file, or
- The available method of the underlying stream returns zero, indicating that further input requests would block. 

if it reads the maximum amount of bytes this means that the request is bigger than the configured maximum and the request should be dropped.

What needs to be improved is the exception handling, but this is a work for another day.

File: src/server/Server.java (hunk 4 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java:163]: > PathFilter is used to mitigate LFI attacks

Shouldn't the user implement it themselves

File: src/lib/Network.java (hunk 11 above)
Comment (Zelakolase) [reply src/lib/Network.java]: The available method is not reliable enough in TLS communication. [Check this discussion I made a year ago](https://coderanch.com/t/747813/java/SSL-TLS-Record-reading-java).

image

File: src/lib/Network.java (hunk 10 above)
Comment (Zelakolase) [reply src/lib/Network.java]: Request **too** big, not major issue

File: src/server/Server.java (hunk 9 above)
Comment (Zelakolase) [reply src/server/Server.java]: That should be noted in the README file

File: src/server/Server.java (hunk 5 above)
Comment (Zelakolase) [reply src/server/Server.java]: ContentTypes should be acquired from MIME.db as well

File: src/server/Server.java (hunk 4 above)
Comment (Zelakolase) [reply src/server/Server.java:163]: Nope, it is a critical issue

File: src/test/SimpleLogin.java (hunk 1 above)
Comment (Zelakolase) [reply src/test/SimpleLogin.java]: Please use JSON class to show how it is being used. I've added the custom header example

File: src/lib/Network.java (hunk 11 above)
Comment (Zelakolase) [reply src/lib/Network.java]: We can sort it out if we experimented sending +16kb post data over tls and see how it is going to read

File: src/test/SimpleLogin.java (hunk 1 above)
Comment (omar-mohamed-khallaf) [reply src/test/SimpleLogin.java]: I don't know how to use it and to be honest I don't like it.

File: src/test/ServeFiles.java (hunk 3 above)
Comment (omar-mohamed-khallaf) [reply src/test/ServeFiles.java]: > Why did you use String wrapper? What if it is a raw file (eg. mp4)?

I think this is resolved by the last commit.

Comment (omar-mohamed-khallaf) [reply src/test/ServeFiles.java]: > Please add JKS comment, in the SimpleLogin class as well

You are free to modify any of the examples.

File: src/server/Server.java (hunk 4 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java:163]: I really don't want to understand it and adapt it to the new interface, can you do it.

File: src/server/Server.java (hunk 5 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java]: I think it's unnecessary, the interface is good for now.
We can modify it in a different version.

File: src/server/Server.java (hunk 6 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java]: After looking into its meaning & usage I think it must be `0` and the user shouldn't modify it.

File: src/server/Server.java (hunk 7 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java]: DOS attacks just need powerful hardware XD

File: src/server/Server.java (hunk 8 above)
Comment (omar-mohamed-khallaf) [reply src/server/Server.java]: No, No one will need to change the port of the server after starting the it, and most of the configurations need to be configured once at the begining.

File: src/lib/Network.java (hunk 11 above)
Comment (omar-mohamed-khallaf) [reply src/lib/Network.java]: I hope the new commit fixes this problem

**Comment by Zelakolase** [comment]
Merging will happen after doing proper tests and implement reading buffer per buffer until reaching end of request or max size, which is earlier

**Comment by Zelakolase** [comment]
Current issues to be addressed before merging:

1. Strange error when running `server.startHttps("src/etc/keystore.jks", "123456");`, Exception in thread "Thread-2" Exception in thread "Thread-3" java.lang.RuntimeException: java.net.BindException: Address already in use. Tested ports: 443, 8443, 16000 with startHttps.

2. Buffer-by-buffer reading till we reach the end of the request OR http.config.ServerConfig.MAX_REQUEST_SIZE_BYTES.

3. (Related to Point.2) When Max sizes in ServerConfig are 5242880 bytes, repeated serial requests can make RAM usage go up to 30MB.

**Comment by Zelakolase** [comment]
Back to work!

**Comment by Zelakolase** [comment]
Issue 1: What is the goal of having a backlog when line 188 in src/Server only considers MAX_CONCURRENT_CONNECTIONS?

**Comment by Zelakolase** [comment]
Issue 2: Please assist in commenting the HttpResponse class according to the Java Convention

**Comment by Zelakolase** [comment]
Continuation of Issue 2: and KeyAttachment Class

**Comment by Zelakolase** [comment]
Continuation of Issue 2: and Server class

**Comment by Zelakolase** [comment]
Issue 3 (Last): HTTPS NIO Impl

**Pull request closed without merging**